networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/DiscoveryNetwork.java

```diff
@@ -121,19 +121,15 @@ private static DiscoveryService createDiscoveryService(
   }
 
   @Override
-  public void stop() {
-    connectionManager
+  public SafeFuture<?> stop() {
+    return connectionManager
         .stop()
         .exceptionally(
             error -> {
               LOG.error("Failed to stop connection manager", error);
               return null;
             })
-        .always(
-            () -> {
-              p2pNetwork.stop();
-              discoveryService.stop().reportExceptions();
-            });
+        .thenCompose((__) -> SafeFuture.allOf(p2pNetwork.stop(), discoveryService.stop()));
```

Comment:
We need to make sure this still runs regardless of whether connectionManager.stop() failed or not.  You actually want a `.handleComposed` method on `SafeFuture` I suspect.  Would be something like:

```
  /**
   * Returns a new CompletionStage that, when this stage completes either normally or exceptionally,
   * is executed with this stage's result and exception as arguments to the supplied function.
   *
   * <p>When this stage is complete, the given function is invoked with the result (or {@code null}
   * if none) and the exception (or {@code null} if none) returning another `CompletionStage`. When
   * that stage completes, the `SafeFuture` returned by this method is completed with the same value
   * or exception.
   *
   * @param fn the function to use to compute another CompletionStage
   * @param <U> the function's return type
   * @return the new SafeFuture
   */
  public <U> SafeFuture<U> handleComposed(
      final BiFunction<? super T, Throwable, CompletionStage<U>> fn) {
    final SafeFuture<U> result = new SafeFuture<>();
    whenComplete(
        (value, error) -> {
          try {
            propagateResult(fn.apply(value, error), result);
          } catch (final Throwable t) {
            result.completeExceptionally(t);
          }
        });
    return result;
  }
```

That's quite a useful method to have around actually.  Would need some extra tests for it.

Reply:
I actually noticed that I could just do instead:
```
  public SafeFuture<?> stop() {
    return SafeFuture.allOf(
            connectionManager.stop(),
            p2pNetwork.stop(),
            discoveryService.stop()
    );
```
Went with this since it's much simpler. Let me know if you don't think that's a good approach.

Reply:
The tests in `DiscoveryNetworkTest` implied that connection manager would need to be stopped before p2pNetwork or discovery service, but after I inspected connectionManager.stop() I don't believe that's necessarily the case. Thus just simply stopping them all together makes sense to me.

Reply:
I don't think that's right.  The connection manager is stopped first so that it stops connecting to new peers and running discovery searches, otherwise you get an awful lot of race conditions.

Reply:
That makes sense. Added `handleComposed` and two unit tests along with it.